libethereum/EthereumCapability.cpp

```diff
@@ -951,6 +940,58 @@ void EthereumCapability::removeSentTransactions(std::vector<h256> const& _txHash
     }
 }
 
+void EthereumCapability::propagateNewBlocks(std::shared_ptr<VerifiedBlocks const> const& _newBlocks)
+{
+    // Safe to call isSyncing() from a non-network thread since the underlying variable is
+    // std::atomic
+    if (_newBlocks->empty() || isSyncing())
+        return;
+
+    m_host->postWork([this, _newBlocks]() {
+        // Verify that we're not too far behind - we perform this check on the network thread to
+        // simplify the synchronization story (since otherwise we'd need to synchronize access to
+        // m_latestBlockSent)
+        auto const latestHash = _newBlocks->back().verified.info.hash();
+        auto const detailsFrom = m_chain.details(m_latestBlockSent);
+        auto const detailsTo = m_chain.details(latestHash);
+        if (detailsFrom.totalDifficulty >= detailsTo.totalDifficulty)
+            return;
+
+        if (diff(detailsFrom.number, detailsTo.number) > c_maxSendNewBlocksCount)
+            return;
+
+        auto const peersWithoutBlock = selectPeers(
+            [&](EthereumPeer const& _peer) { return !_peer.isBlockKnown(latestHash); });
+
+        auto const peersToSendNumber =
+            std::max<std::size_t>(c_minBlockBroadcastPeers, std::sqrt(m_peers.size()));
+
+        std::vector<NodeID> const peersToSend = randomPeers(peersWithoutBlock, peersToSendNumber);
+        for (NodeID const& peerID : peersToSend)
+            for (auto const& b : *_newBlocks)
+            {
+                RLPStream ts;
+                m_host->prep(peerID, name(), ts, NewBlockPacket, 2)
+                    .appendRaw(b.blockData, 1)
+                    .append(b.verified.info.difficulty());
+
+                auto itPeer = m_peers.find(peerID);
+                if (itPeer != m_peers.end())
+                {
+                    m_host->sealAndSend(peerID, ts);
+                    // We don't want to send new block hashes to these same peers
+                    itPeer->second.markBlockAsKnown(b.verified.info.hash());
```

Comment:
There was `clearKnownBlocks()` instead of this in the old code, do you have explanation for this change?

Reply:
Yes, this change was intentional - in the old code (`EthereumCapability::maintainBlocks`), we used the "known block" list in each peer to avoid sending new blocks and hashes to peers who already know about the block i.e. peers from whom we received the new block or block hash. We would clear this list after each iteration of sending out new blocks / hashes because we no longer need to track this information for that batch of blocks (since they've already been propagated and will not be re-propagated). 

In the new code, we can't clear this information after propagating new blocks since we still need to use it when we send out new block hashes (in `EthereumCapability::maintainBlockHashes`) to avoid sending hashes to peers who already know about the block.

One thing I'm not sure about is why - in the original code in `EthereumCapability::maintainBlocks` - we cleared a peer's know blocks list after sending them each block, rather than waiting until we've propagated all of the blocks to them:https://github.com/ethereum/aleth/blob/7e6095dc3d6c9cefac4b0867f60d06f955ab7c7e/libethereum/EthereumCapability.cpp#L539-L553 

 I don't see any benefit to doing this or any downside from not doing this...we only use the known blocks list to determine which peers to send blocks to in `maintainBlocks `so it doesn't have anything to do with synchronizing actions across different threads.